File: cal.py
# 1
import pandas as pd 
import numpy as np  
# Importing the dataset
food = pd.read_csv("C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Linear Regression\\calories_consumed.csv")
food=food.rename(columns={'Weight gained (grams)':'weight_gained','Calories Consumed':'calories'})
# Performing Explorory data analysis
food.describe()
# Graphical Representation
import matplotlib.pyplot as plt # mostly used for visualization purposes 
# Barplot
plt.bar(height = food.calories, x = np.arange(1, 15, 1))
# Histogram
plt.hist(food.calories) 
# Boxplot
plt.boxplot(food.calories) 
plt.bar(height = food.weight_gained, x = np.arange(1, 15, 1))
plt.hist(food.weight_gained) #histogram
plt.boxplot(food.weight_gained) #boxplot
# Scatter plot
plt.scatter(x = food['weight_gained'], y = food['calories'], color = 'green') 
# correlation
np.corrcoef(food.weight_gained, food.calories) 
# Covariance
cov_output = np.cov(food.weight_gained, food.calories)[0, 1]
cov_output
food.cov()
import statsmodels.formula.api as smf
# Simple Linear Regression
model = smf.ols('calories ~ weight_gained', data = food).fit()
model.summary()
pred1 = model.predict(pd.DataFrame(food['weight_gained']))
# Regression Line
plt.scatter(food.weight_gained, food.calories)
plt.plot(food.weight_gained, pred1, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res1 = food.calories - pred1
res_sqr1 = res1 * res1
mse1 = np.mean(res_sqr1)
rmse1 = np.sqrt(mse1)
rmse1
# Model building on Transformed Data
# Log Transformation
# x = log(weight_gained); y = at
plt.scatter(x = np.log(food['weight_gained']), y = food['calories'], color = 'brown')
np.corrcoef(np.log(food.weight_gained), food.calories) #correlation
model2 = smf.ols('calories ~ np.log(weight_gained)', data = food).fit()
model2.summary()
pred2 = model2.predict(pd.DataFrame(food['weight_gained']))
# Regression Line
plt.scatter(np.log(food.weight_gained), food.calories)
plt.plot(np.log(food.weight_gained), pred2, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res2 = food.calories - pred2
res_sqr2 = res2 * res2
mse2 = np.mean(res_sqr2)
rmse2 = np.sqrt(mse2)
rmse2
# Exponential transformation
# x = weight_gained; y = log(at)
plt.scatter(x = food['weight_gained'], y = np.log(food['calories']), color = 'orange')
np.corrcoef(food.weight_gained, np.log(food.calories)) #correlation
model3 = smf.ols('np.log(calories) ~ weight_gained', data = food).fit()
model3.summary()
pred3 = model3.predict(pd.DataFrame(food['weight_gained']))
pred3_at = np.exp(pred3)
pred3_at
# Regression Line
plt.scatter(food.weight_gained, np.log(food.calories))
plt.plot(food.weight_gained, pred3, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res3 = food.calories - pred3_at
res_sqr3 = res3 * res3
mse3 = np.mean(res_sqr3)
rmse3 = np.sqrt(mse3)
rmse3
# Polynomial transformation
# x = weight_gained; x^2 = weight_gained*weight_gained; y = log(at)
model4 = smf.ols('np.log(calories) ~ weight_gained+ I(weight_gained*weight_gained)', data = food).fit()
model4.summary()
pred4 = model4.predict(pd.DataFrame(food))
pred4_at = np.exp(pred4)
pred4_at
# Regression line
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 2)
X = food.iloc[:, 0:1].values
X_poly = poly_reg.fit_transform(X)
# y = food.iloc[:, 1].values
plt.scatter(food.weight_gained, np.log(food.calories))
plt.plot(X, pred4, color = 'red')
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res4 = food.calories - pred4_at
res_sqr4 = res4 * res4
mse4 = np.mean(res_sqr4)
rmse4 = np.sqrt(mse4)
rmse4
# Choose the best model using RMSE
data = {"MODEL":pd.Series(["SLR", "Log model", "Exp model", "Poly model"]), "RMSE":pd.Series([rmse1, rmse2, rmse3, rmse4])}
table_rmse = pd.DataFrame(data)
table_rmse
# The best model
from sklearn.model_selection import train_test_split
train, test = train_test_split(food, test_size = 0.3)
finalmodel = smf.ols('np.log(calories) ~ weight_gained + I(weight_gained*weight_gained)', data = train).fit()
finalmodel.summary()
# Predict on test data
test_pred = finalmodel.predict(pd.DataFrame(test))
pred_test_calories = np.exp(test_pred)
pred_test_calories
# Model Evaluation on Test data
test_res = test.calories - pred_test_calories
test_sqrs = test_res * test_res
test_mse = np.mean(test_sqrs)
test_rmse = np.sqrt(test_mse)
test_rmse
# Prediction on train data
train_pred = finalmodel.predict(pd.DataFrame(train))
pred_train_calories = np.exp(train_pred)
pred_train_calories
# Model Evaluation on train data
train_res = train.calories - pred_train_calories
train_sqrs = train_res * train_res
train_mse = np.mean(train_sqrs)
train_rmse = np.sqrt(train_mse)
train_rmse

File: cal.R
# 1
library(readr)
# Imporing the dataset
cal <- read.csv('C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Linear Regression\\calories_consumed.csv', header = T)
View(cal)
# Performing EDA
summary(cal)
install.packages("Hmisc")
library(Hmisc)
describe(cal)
library("lattice") 
# Graphical exploration
# Dotplot
dotplot(cal$Weight.gained..grams., main = "Dot Plot of Waist Circumferences")
dotplot(cal$Calories.Consumed, main = "Dot Plot of Adipose Tissue Areas")
# Boxplot
boxplot(cal$Weight.gained..grams, col = "dodgerblue4")
boxplot(cal$Calories.Consumed, col = "red", horizontal = T)
# Histogram
hist(cal$Weight.gained..grams)
hist(cal$Calories.Consumed)
# Normal QQ plot
qqnorm(cal$Weight.gained..grams)
qqline(cal$Weight.gained..grams)
qqnorm(cal$Calories.Consumed)
qqline(cal$Calories.Consumed)
hist(cal$Weight.gained..grams, prob = TRUE)          # prob = TRUE for probabilities not counts
lines(density(cal$Weight.gained..grams))             # add a density estimate with defaults
lines(density(cal$Weight.gained..grams, adjust = 2), lty = "dotted")   # add another "smoother" density

hist(cal$Calories.Consumed, prob = TRUE)          # prob = TRUE for probabilities not counts
lines(density(cal$Calories.Consumed))             # add a density estimate with defaults
lines(density(cal$Calories.Consumed, adjust = 2), lty = "dotted")   # add another "smoother" density
# Bi-variate analysis
# Scatter plot
plot(cal$Weight.gained..grams,cal$Calories.Consumed, main = "Scatter Plot", col = "Dodgerblue4", col.main = "Dodgerblue4", col.lab = "Dodgerblue4", xlab = "Weight.gained..grams", ylab = "Calories.Consumed", pch = 20)

# Correlation Coefficient
cor(cal$Weight.gained..grams, cal$Calories.Consumed)
# Co-variance
cov(cal$Weight.gained..grams,cal$Calories.Consumed)

# Linear Regression model
reg <- lm(cal$Calories.Consumed ~ cal$Weight.gained..grams, data = cal) # Y ~ X
summary(reg)
confint(reg, level = 0.95)
pred <- predict(reg, interval = "predict")
pred <- as.data.frame(pred)
View(pred)

# gg plot for adding Regression line for data
library(ggplot2)
ggplot(data = cal, aes(cal$Weight.gained..grams, cal$Calories.Consumed) ) + geom_point() + stat_smooth(method = lm, formula = y ~ x)

# Evaluation the model for fitness 
cor(pred$fit, cal$Calories.Consumed)
reg$residuals
rmse <- sqrt(mean(reg$residuals^2))
rmse

# Transformation Techniques
# input = log(x); output = y
plot(log(cal$Weight.gained..grams), cal$Calories.Consumed)
cor(log(cal$Weight.gained..grams), cal$Calories.Consumed)
reg_log <- lm(cal$Calories.Consumed ~ log(cal$Weight.gained..grams), data = cal)
summary(reg_log)
confint(reg_log,level = 0.95)
pred <- predict(reg_log, interval = "predict")
pred <- as.data.frame(pred)
cor(pred$fit, cal$Calories.Consumed)
rmse <- sqrt(mean(reg_log$residuals^2))
rmse
# Regression line for data
ggplot(data = cal, aes(log(cal$Weight.gained..grams), cal$Calories.Consumed) ) + geom_point() + stat_smooth(method = lm, formula = y ~ log(x))

# Log transformation applied on 'y'
# input = x; output = log(y)
plot(cal$Weight.gained..grams, log(cal$Calories.Consumed))
cor(cal$Weight.gained..grams, log(cal$Calories.Consumed))
reg_log1 <- lm(log(cal$Calories.Consumed) ~ cal$Weight.gained..grams, data = cal)
summary(reg_log1)
predlog <- predict(reg_log1, interval = "predict")
predlog <- as.data.frame(predlog)
reg_log1$residuals
sqrt(mean(reg_log1$residuals^2))
# Exponential function
pred <- exp(predlog)  
pred <- as.data.frame(pred)
cor(pred$fit, cal$Calories.Consumed)
res_log1 = cal$Calories.Consumed - pred$fit
rmse <- sqrt(mean(res_log1^2))
rmse
# Regression line for data
ggplot(data = cal, aes(x = cal$Weight.gained..grams, y = log(cal$Calories.Consumed))) + geom_point(color = 'blue') + geom_line(color = 'red', data = cal, aes(x = cal$Weight.gained..grams, y = predlog$fit))
# Non-linear models = Polynomial models
# input = x & x^2 (2-degree) and output = log(y)
reg2 <- lm(log(cal$Calories.Consumed) ~ cal$Weight.gained..grams + I(cal$Weight.gained..grams*cal$Weight.gained..grams), data = cal)
summary(reg2)
predlog <- predict(reg2, interval = "predict")
pred <- exp(predlog)
pred <- as.data.frame(pred)
cor(pred$fit, cal$Calories.Consumed)
res2 = cal$Calories.Consumed - pred$fit
rmse <- sqrt(mean(res2^2))
rmse
# Regression line for data
ggplot(data = cal, aes(cal$Weight.gained..grams, log(cal$Calories.Consumed)) ) + geom_point() + stat_smooth(method = lm, formula = y ~ poly(x, 2, raw = TRUE))

# Data Partition
train <- cal[1:7, ]
test <- cal[8:14, ]
log(train$Calories.Consumed)
log(test$Calories.Consumed)
plot(train$Weight.gained..grams., log(train$Calories.Consumed))
plot(test$Weight.gained..grams., log(test$Calories.Consumed))
model <- lm(log(cal$Calories.Consumed) ~ cal$Weight.gained..grams + I(cal$Weight.gained..grams * cal$Weight.gained..grams), data = train)
summary(model)
confint(model,level=0.95)
log_res <- predict(model,interval = "confidence", newdata = test)
# converting log values to original values
predict_original <- exp(log_res) 
predict_original <- as.data.frame(predict_original)
test_error <- test$Calories.Consumed- predict_original$fit # calculate error/residual
test_error
test_rmse <- sqrt(mean(test_error^2))
test_rmse
log_res_train <- predict(model, interval = "confidence", newdata = train)
# converting log values to original values
predict_original_train <- exp(log_res_train) 
predict_original_train <- as.data.frame(predict_original_train)
train_error <- train$Calories.Consumed - predict_original_train$fit # calculate error/residual
train_error
train_rmse <- sqrt(mean(train_error^2))
train_rmse

